## Problem Statement

### Business Context

The healthcare industry is rapidly evolving, with professionals facing increasing challenges in managing vast volumes of medical data while delivering accurate and timely diagnoses. The need for quick access to comprehensive, reliable, and up-to-date medical knowledge is critical for improving patient outcomes and ensuring informed decision-making in a fast-paced environment.

Healthcare professionals often encounter information overload, struggling to sift through extensive research and data to create accurate diagnoses and treatment plans. This challenge is amplified by the need for efficiency, particularly in emergencies, where time-sensitive decisions are vital. Furthermore, access to trusted, current medical information from renowned manuals and research papers is essential for maintaining high standards of care.

To address these challenges, healthcare centers can focus on integrating systems that streamline access to medical knowledge, provide tools to support quick decision-making, and enhance efficiency. Leveraging centralized knowledge platforms and ensuring healthcare providers have continuous access to reliable resources can significantly improve patient care and operational effectiveness.

**Common Questions to Answer**

**1. Diagnostic Assistance**: "What are the common symptoms and treatments for pulmonary embolism?"

**2. Drug Information**: "Can you provide the trade names of medications used for treating hypertension?"

**3. Treatment Plans**: "What are the first-line options and alternatives for managing rheumatoid arthritis?"

**4. Specialty Knowledge**: "What are the diagnostic steps for suspected endocrine disorders?"

**5. Critical Care Protocols**: "What is the protocol for managing sepsis in a critical care unit?"

### Objective

As an AI specialist, your task is to develop a RAG-based AI solution using renowned medical manuals to address healthcare challenges. The objective is to **understand** issues like information overload, **apply** AI techniques to streamline decision-making, **analyze** its impact on diagnostics and patient outcomes, **evaluate** its potential to standardize care practices, and **create** a functional prototype demonstrating its feasibility and effectiveness.

### Data Description

The **Merck Manuals** are medical references published by the American pharmaceutical company Merck & Co., that cover a wide range of medical topics, including disorders, tests, diagnoses, and drugs. The manuals have been published since 1899, when Merck & Co. was still a subsidiary of the German company Merck.

The manual is provided as a PDF with over 4,000 pages divided into 23 sections.

## Installing and Importing Necessary Libraries and Dependencies

In [1]:
# Installation for GPU llama-cpp-python
# uncomment and run the following code in case GPU is being used
!CMAKE_ARGS="-DLLAMA_CUBLAS=on" FORCE_CMAKE=1 pip install llama-cpp-python==0.2.28 --force-reinstall --no-cache-dir -q

# Installation for CPU llama-cpp-python
# uncomment and run the following code in case GPU is not being used
# !CMAKE_ARGS="-DLLAMA_CUBLAS=off" FORCE_CMAKE=1 pip install llama-cpp-python==0.2.28 --force-reinstall --no-cache-dir -q

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 9.4/9.4 MB 31.4 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Installing backend dependencies ... done
  Preparing metadata (pyproject.toml) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 45.5/45.5 kB 241.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 16.6/16.6 MB 365.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 44.6/44.6 kB 274.3 MB/s eta 0:00:00
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
numba 0.60.0 requires numpy<2.1,>=1.22, but you have numpy 2.4.6 which is incompatible.


**Note**:
- After running the above cell, kindly restart the runtime (for Google Colab) or notebook kernel (for Jupyter Notebook), and run all cells sequentially from the next cell.
- On executing the above line of code, you might see a warning regarding package dependencies. This error message can be ignored as the above code ensures that all necessary libraries and their dependencies are maintained to successfully execute the code in ***this notebook***.

In [1]:
# For installing the libraries & downloading models from HF Hub
!pip install huggingface_hub==0.35.3 pandas==2.2.2 tiktoken==0.12.0 pymupdf==1.26.5 langchain==0.3.27 langchain-community==0.3.31 chromadb==1.1.1 sentence-transformers==5.1.1 numpy==2.3.3 -q

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 62.1/62.1 kB 266.2 kB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 52.0/52.0 kB 6.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 44.0/44.0 kB 5.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 564.3/564.3 kB 31.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 24.1/24.1 MB 103.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.0/1.0 MB 73.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.5/2.5 MB 49.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 19.9/19.9 MB 99.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 486.6/486.6 kB 46.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 16.6/16.6 MB 24.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 278.2/278.2 kB 32.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.7/4.7 MB 129.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 46

**Note**:
- After running the above cell, kindly restart the runtime (for Google Colab) or notebook kernel (for Jupyter Notebook), and run all cells sequentially from the next cell.
- On executing the above line of code, you might see a warning regarding package dependencies. This error message can be ignored as the above code ensures that all necessary libraries and their dependencies are maintained to successfully execute the code in ***this notebook***.

In [1]:
#Libraries for processing dataframes,text
import json,os
import tiktoken
import pandas as pd

#Libraries for Loading Data, Chunking, Embedding, and Vector Databases
from langchain.text_splitter import RecursiveCharacterTextSplitter
from langchain_community.document_loaders import PyMuPDFLoader
from langchain_community.embeddings.sentence_transformer import SentenceTransformerEmbeddings
from langchain_community.vectorstores import Chroma

#Libraries for downloading and loading the llm
from huggingface_hub import hf_hub_download
from llama_cpp import Llama

## Question Answering using LLM

#### Downloading and Loading the model

In [2]:
## Model configuration

# Using hf_hub_download to download a model from the Hugging Face model hub
# The repo_id parameter specifies the model name or path in the Hugging Face repository
# The filename parameter specifies the name of the file to download
model_name_or_path = "TheBloke/Llama-2-13B-chat-GGUF"
model_basename = "llama-2-13b-chat.Q5_K_M.gguf"
model_path = hf_hub_download(
    repo_id=model_name_or_path,
    filename=model_basename
    )

/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


llama-2-13b-chat.Q5_K_M.gguf:   0%|          | 0.00/9.23G [00:00<?, ?B/s]

In [3]:
#Initializing the Llama object
lcpp_llm = Llama(
    model_path=model_path,
    n_threads=2,  # CPU cores
    n_batch=256,  # Should be between 1 and n_ctx, consider the amount of VRAM in your GPU.
    n_gpu_layers=32,  # Change this value based on your model and your GPU VRAM pool.
    n_ctx=4096,  # Context window
)

AVX = 1 | AVX_VNNI = 0 | AVX2 = 1 | AVX512 = 1 | AVX512_VBMI = 0 | AVX512_VNNI = 1 | FMA = 1 | NEON = 0 | ARM_FMA = 0 | F16C = 1 | FP16_VA = 0 | WASM_SIMD = 0 | BLAS = 1 | SSE3 = 1 | SSSE3 = 1 | VSX = 0 | 


#### Response

In [4]:
# Defining Model Response Parameters
# function to generate, process, and return the response from the LLM
def generate_llama_response(user_prompt):

    # System message
    system_message = """
    [INST]<<SYS>> Respond to the user question based on the user prompt<</SYS>>[/INST]
    """

    # Combine user_prompt and system_message to create the prompt
    prompt = f"{user_prompt}\n{system_message}"

    # Generate a response from the LLaMA model
    response = lcpp_llm(
        prompt=prompt,
        max_tokens=128,
        temperature=0.01,
        top_p=0.95,
        repeat_penalty=1.2,
        top_k=50,
        stop=['INST'],
        echo=False
    )

    # Extract and return the response text
    response_text = response["choices"][0]["text"]
    return response_text

### Query 1: What is the protocol for managing sepsis in a critical care unit?

In [5]:
user_prompt = "What is the protocol for managing sepsis in a critical care unit?"
response = generate_llama_response(user_prompt)
print(response)

 The management of sepsis in a critical care unit (CCU) follows a standardized protocol that includes early recognition, aggressive fluid resuscitation, antibiotics administration, and close monitoring. Here are the key components of the protocol:

1. Early Recognition: The CCU team should be vigilant in identifying patients at risk for sepsis, using criteria such as SIRS (Systemic Inflammatory Response Syndrome) or QSOFA (Quick Sequential Organ Failure Assessment).
2. Aggressive Fluid Resusc


### Query 2: What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?

In [ ]:
user_prompt = "What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?"
response = generate_llama_response(user_prompt)
print(response)

Llama.generate: prefix-match hit


 Sure, I'd be happy to help! Appendicitis is a medical emergency that requires prompt treatment. Here are some common symptoms of appendicitis:

1. Severe pain in the abdomen, usually starting near the belly button and then moving to the lower right side of the abdomen. The pain may be constant or it may come and go.
2. Nausea and vomiting.
3. Loss of appetite.
4. Fever (usually above 101°F).
5. Abdominal tenderness and guarding


### Query 3: What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?

In [ ]:
user_prompt = "What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?"
response = generate_llama_response(user_prompt)
print(response)

Llama.generate: prefix-match hit


 Sudden patchy hair loss, also known as alopecia areata, can be a distressing condition that affects millions of people worldwide. While there is no definitive cure for this condition, there are several effective treatments and solutions that can help promote hair growth and reduce the size of bald spots.

     One of the most common treatment options for alopecia areata is topical corticosteroids, which can be applied directly to the affected area to suppress the immune system and promote hair growth. Another effective treatment option is anthralin, a medication that is applied to


### Query 4:  What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?

In [ ]:
user_prompt = "What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?"
response = generate_llama_response(user_prompt)
print(response)

Llama.generate: prefix-match hit


 Treatment options for a person with a physical injury to brain tissue depend on several factors such as the location and severity of the injury, age, overall health status, and personal preferences. Here are some common treatment approaches: 1) Medications: To manage symptoms like pain, anxiety, or depression; 2) Rehabilitation therapy: Physical, occupational, speech, and cognitive therapies to improve function and independence; 3) Surgery: May be necessary to relieve pressure on the brain, repair damaged blood vessels, or remove a tumor;


### Query 5: What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?

In [ ]:
user_prompt = "What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?"
response = generate_llama_response(user_prompt)
print(response)

Llama.generate: prefix-match hit


 Sure, I'd be happy to help! If someone has fractured their leg during a hiking trip, there are several necessary precautions and treatment steps that should be taken to ensure proper care and recovery. Here are some considerations:

1. Stop activity: The first step is to stop any further activity or movement that may exacerbate the injury. This includes ceasing all hiking-related activities and returning to a safe location.
2. Assess the severity of the fracture: It's important to assess the severity of the fracture and determine if


#### Observations :

  All the responses seem to be generic and align with the query , but are restricted and limited due to the number of max_tokens specified as 128.

## Question Answering using LLM with Prompt Engineering

In [11]:
import tiktoken

# function to generate, process, and return the response from the LLM
def generate_llama_response(user_prompt):

    # System message
    system_message = """
    [INST]<<SYS>> Respond to the user question based on the user prompt<</SYS>>[/INST]
    """

    # Combine user_prompt and system_message to create the prompt
    prompt = f"{user_prompt}\n{system_message}"

    # Generate a response from the LLaMA model
    response = lcpp_llm(
        prompt=prompt,
        max_tokens=1024,
        temperature=0.01,
        top_p=0.95,
        repeat_penalty=1.2,
        top_k=50,
        stop=['INST'],
        echo=False
    )

    # Extract and return the response text
    response_text = response["choices"][0]["text"]

    # Get the number of completion tokens
    completion_tokens = response["usage"]["completion_tokens"]
    print(f"Completion Tokens: {completion_tokens}")

    return response_text

### Query 1: What is the protocol for managing sepsis in a critical care unit?

#### 1. Prompt Engineering type 1 -Providing "clear and specific" instructions to lead better results  

In [ ]:
user_prompt = "Create a comprehensive and detailed instructions to be carefully followed by a medical team for managing sepsis in a critical care unit, which is a life threatening medical emergency for any patient."
response = generate_llama_response(user_prompt)
print(response)

Llama.generate: prefix-match hit


Completion Tokens: 949
 Sure, I can help you with that! Here are comprehensive and detailed instructions for managing sepsis in a critical care unit:

I. Identification of Sepsis

* Use a systematic approach to identify patients at risk for sepsis, such as the Surviving Sepsis Campaign (SSC) bundle or the Quick Sequential Organ Failure Assessment (qSOFA).
* Monitor vital signs and laboratory values regularly, especially temperature, blood pressure, tissue perfusion, and organ function.
* Use a standardized sepsis protocol to guide assessment and management of patients with suspected or confirmed sepsis.

II. Initial Management of Sepsis

* Administer broad-spectrum antibiotics within the first hour of recognition of sepsis, preferably before performing any diagnostic tests.
* Use a vasopressor (such as norepinephrine or dopamine) to maintain mean arterial pressure ≥65 mmHg and serum lactate level <2 mmol/L.
* Provide oxygen therapy as needed, with the goal of maintaining SpO2 ≥94%.
* C

#### Observations -

   The response given is very clear with specific details. It is according to the prompt given and elaborate as the max_tokens is increased. The response is given in 949 tokens.

### Query 2: What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?

####2. Prompt Engineering type 2 -


Prompts separated by delimiters for 2 different sections for the query . In the second section, conditional prompting is applied based on a certain criterion.

In [ ]:
user_prompt = '''Elaborate on the common symptoms for appendicitis and it's cure.

 If appendicitis can be cured by medicine respond by listing the names of the medicines to give to the patients along with specific instructions
 Else - mention what surgical procedure should be followed to treat it.

'''

response = generate_llama_response(user_prompt)

print(response)

Llama.generate: prefix-match hit


Completion Tokens: 556
 Sure, I'd be happy to help! Appendicitis is a medical emergency that requires immediate attention. Here are some common symptoms of appendicitis:

1. Severe pain in the abdomen, usually starting near the belly button and then moving to the lower right side of the abdomen. The pain may be constant or it may come and go.
2. Nausea and vomiting.
3. Loss of appetite.
4. Fever.
5. Abdominal tenderness and guarding (muscle tension) in the affected area.
6. Abdominal swelling.
7. Diarrhea or constipation.
8. Inability to pass gas.

If you suspect that you or someone else may have appendicitis, it is important to seek medical attention immediately. Appendicitis can be treated with antibiotics and fluids through an IV, but if the appendix ruptures, surgery is necessary to remove the inflamed tissue.

The medications used to treat appendicitis are:

1. Antibiotics such as Cefotaxime or Metronidazole to treat any secondary infections.
2. Pain management medication such as 

#### Observations -

   The response given is according to the prompt given. The response is given in 546 tokens.
   It has answered both the sections of the prompt specifically.

   The first part of the prompt asking for Symptoms of the Appdicitis is given in specific 8 steps or bullet points.

   The conditional statement ( IF - Else) is also addressed correctly giving information about medication cure and if not through cure by medicine, what surgical procedure should be done.

### Query 3: What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?

####3. Prompt Engineering type 3 -



Prompts to give structured outputs in the form of JSON


In [12]:
user_prompt = '''  Give me the top three effective treatments or solutions for addressing sudden patchy hair loss

The output should be in the form of a JSON with

1. Type of hair loss  (as string)
2. Causes behind the hairloss  (as string)
3. Treatment or solutions  (as string)

'''
response = generate_llama_response(user_prompt)
print(response)

Llama.generate: prefix-match hit


Completion Tokens: 436
 Sure, here are the top three effective treatments or solutions for addressing sudden patchy hair loss:

{
"1": {
"Type of Hair Loss": "Patchy Hair Loss",
"Causes Behind Hair Loss": "Genetics, Hormonal Imbalance, Autoimmune Disorders, Infections, and Nutritional Deficiencies",
"Treatment or Solutions": "Medications such as Minoxidil (Rogaine) and Finasteride (Propecia), Low-Level Laser Therapy (LLLT), Platelet-Rich Plasma (PRP) Therapy, Dietary Changes, and Hair Transplantation"
},

"2": {
"Type of Hair Loss": "Alopecia Areata",
"Causes Behind Hair Loss": "Autoimmune Disorder that Attacks Healthy Hair Follicles",
"Treatment or Solutions": "Topical Corticosteroids, Immunomodulators such as PUVA and Psoralen + UV-A, Intralesional Therapy with Steroids or Chemicals, Systemic Medications like Methotrexate and Cyclosporine, Wigs and Hairpieces"
},

"3": {
"Type of Hair Loss": "Telogen Effluvium",
"Causes Behind Hair Loss": "Physical or Emotional Stress, Hormonal Imbal

#### Observations -

   The JSON format response is according to the given prompt. The response is given in 436 tokens. The model has correctly given 3 treatments and segregated the type, causes and treatments for sudden , patchy Hairloss.
   

### Query 4:  What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?

####4. Prompt Engineering type 4 -



More specific and elaborate instructions with max_tokens increased to 1024.


In [13]:
user_prompt ='''  As a medical assistant elucidate the different treatments that needs to be done for a person who has sustained a physical injury to brain tissues,
 resulting in temporary or permanent impairment of brain function?

 Kindly give the response with easier readability separated by a new line.
'''

response = generate_llama_response(user_prompt)
print (response)

Llama.generate: prefix-match hit


Completion Tokens: 667
 Sure, I'd be happy to help! Treatment for brain injuries can vary depending on the severity and location of the injury, as well as the individual needs of the patient. Here are some common treatments that may be used for a person who has sustained a physical injury to brain tissues:

1. Medications: Depending on the type and severity of the injury, medications such as pain relievers, anti-seizure drugs, or antidepressants may be prescribed to help manage symptoms.
2. Rehabilitation therapy: Physical, occupational, and speech therapy can help patients regain lost function and improve their quality of life. This type of therapy may include exercises to strengthen weakened muscles, activities to improve cognitive function, and techniques to enhance communication skills.
3. Surgery: In some cases, surgery may be necessary to relieve pressure on the brain or repair damaged tissue. For example, a patient with a severe concussion may require surgical intervention to re

#### Observations -

   The JSON format response is according to the given prompt. The response is given in 436 tokens. The model has correctly given 3 treatments and segregated the type, causes and treatments for sudden , patchy Hairloss.
   

### Query 5: What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?

####5. Prompt Engineering type 5 -



Prompt to give responses and instructions step-wise.


In [ ]:
user_prompt = '''Please specify in  multiple steps the necessary precautions and treatments for a person who has fractured their leg during a hiking trip. Please specify in the last step what should be considered for their care and recovery?
'''

response = generate_llama_response(user_prompt)

print(response)

Llama.generate: prefix-match hit


Completion Tokens: 647
 Sure, I'd be happy to help! Here are the necessary precautions and treatments for a person who has fractured their leg during a hiking trip in multiple steps:

Step 1: Stop all activity and assess the injury
-----------------------------------------------

If the person suspects that they have fractured their leg, they should stop all activity immediately and assess the injury. They should check for any deformity or abnormal alignment of the limb, and look for signs of bleeding or swelling. If there is significant pain or difficulty moving the limb, it is likely that a fracture has occurred.

Step 2: Immobilize the affected leg
-------------------------------

To immobilize the affected leg, the person should use a splint or brace to keep the bone in place and prevent further movement. This can be done using materials found in nature, such as sticks and leaves, or by creating a makeshift splint from their clothing or gear. The goal is to keep the limb stable and

#### Observations -

   The model has given step-wise responses according to the given prompt. The response is given in 647 tokens. The step-wise information and instructions are meaningful and addresses the issues for the injury specified .
   

## Data Preparation for RAG

In [5]:
#Libraries for processing dataframes,text
import json,os
import tiktoken
import pandas as pd

#Libraries for Loading Data, Chunking, Embedding, and Vector Databases
from langchain.text_splitter import RecursiveCharacterTextSplitter
from langchain_community.document_loaders import PyMuPDFLoader
from langchain_community.embeddings.sentence_transformer import SentenceTransformerEmbeddings
from langchain_community.vectorstores import Chroma

### Loading the Data

In [6]:
# uncomment and run the below code snippets if the dataset is present in the Google Drive
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [7]:
manual_pdf_path = "/content/drive/MyDrive/GL/NLP-GenAI/Project/Project/medical_diagnosis_manual.pdf"

In [8]:
pdf_loader = PyMuPDFLoader(manual_pdf_path)

In [9]:
manual = pdf_loader.load()

### Data Overview

#### Checking the first 5 pages

In [10]:
for i in range(5):
    print(f"Page Number : {i+1}",end="\n")
    print(manual[i].page_content,end="\n")

Page Number : 1
bhuvaneshwari_s@yahoo.com
AT5WEX37OM
This file is meant for personal use by bhuvaneshwari_s@yahoo.com only.
Sharing or publishing the contents in part or full is liable for legal action.
Page Number : 2
bhuvaneshwari_s@yahoo.com
AT5WEX37OM
This file is meant for personal use by bhuvaneshwari_s@yahoo.com only.
Sharing or publishing the contents in part or full is liable for legal action.
Page Number : 3
Table of Contents
1
Front    ................................................................................................................................................................................................................
1
Cover    .......................................................................................................................................................................................................
2
Front Matter    ...............................................................................................................

##### Observations - The first 5 pages are loaded correctly from the manual.

#### Checking the number of pages

In [11]:
len(manual)

4114

The manual has 4114 pages

### Data Chunking

In [30]:
text_splitter = RecursiveCharacterTextSplitter.from_tiktoken_encoder(
    encoding_name='cl100k_base',
    chunk_size=512,
    chunk_overlap= 20
)

The given code initializes a **RecursiveCharacterTextSplitter** to split the text into manageable chunks for embedding and retrieval. Here's a breakdown:

- `RecursiveCharacterTextSplitter.from_tiktoken_encoder(...)`: Uses **TikToken encoding** to properly handle token-based splitting.
- `encoding_name='cl100k_base'`: Specifies the **TikToken encoding** (used by OpenAI models like GPT-4 and GPT-3.5).
- `chunk_size=512`: Each text chunk will have a maximum of **512 tokens**.
- `chunk_overlap=20`: Ensures **overlapping** of 20 tokens between consecutive chunks to preserve context.

This approach ensures that text is split **intelligently** while maintaining **semantic meaning** for better retrieval and embeddings.

In [31]:
document_chunks = pdf_loader.load_and_split(text_splitter)

In [32]:
len(document_chunks)

8492

In [33]:
document_chunks[2].page_content

'Table of Contents\n1\nFront    ................................................................................................................................................................................................................\n1\nCover    .......................................................................................................................................................................................................\n2\nFront Matter    ...........................................................................................................................................................................................\n53\n1 - Nutritional Disorders    ...............................................................................................................................................................\n53\nChapter 1. Nutrition: General Considerations    ...........................................................................................

In [34]:
document_chunks[3].page_content

"Chapter 24. Testing for Hepatic & Biliary Disorders    ......................................................................................................\n305\nChapter 25. Drugs & the Liver    ................................................................................................................................................\n308\nChapter 26. Alcoholic Liver Disease    ....................................................................................................................................\n314\nChapter 27. Fibrosis & Cirrhosis    ............................................................................................................................................\n322\nChapter 28. Hepatitis    ..................................................................................................................................................................\n333\nChapter 29. Vascular Disorders of the Liver    .................................................

In [35]:
document_chunks[100].page_content

'compatible with good health). Vegans may develop vitamin B12 deficiency unless they consume yeast\nextracts or Asian-style fermented foods. Their intake of Ca, iron, and zinc also tends to be low. A fruit-only\ndiet is not recommended because it is deficient in protein, Na, and many micronutrients.\nFad diets: Some fad diets result in vitamin, mineral, and protein deficiencies; cardiac, renal, and\nmetabolic disorders; and sometimes death. Very low calorie diets (< 400 kcal/day) cannot sustain health\nfor long.\nDrugs and nutritional supplements: Many drugs (eg, appetite suppressants, digoxin) decrease\nappetite; others impair nutrient absorption or metabolism. Some drugs (eg, stimulants) have catabolic\neffects. Certain drugs can impair absorption of many nutrients; eg, anticonvulsants can impair absorption\nof vitamins.\nAlcohol or drug dependency: Patients with alcohol or drug dependency may neglect their nutritional\nneeds. Absorption and metabolism of nutrients may also be impair

In [36]:
document_chunks[101].page_content

"Diagnosis is based on results of medical and diet histories, physical examination, body composition\nanalysis (see p. 58), and selected laboratory tests.\nHistory: History should include questions about dietary intake (see\nFig. 2-1), recent changes in weight, and risk factors for undernutrition, including drug and alcohol use.\nUnintentional loss of ≥ 10% of usual body weight during a 3-mo period indicates a high probability of\nundernutrition. Social history should include questions about whether money is available for food and\nwhether the patient can shop and cook.\nReview of systems should focus on symptoms of nutritional deficiencies (see Table 2-1). For example,\nimpaired night vision may indicate vitamin A deficiency.\nPhysical examination: Physical examination should include measurement of height and weight,\ninspection of body fat distribution, and anthropometric measurements of lean body mass. Body mass\nindex (BMI = weight(kg)/height(m)2) adjusts weight for height (see\nTa

In [ ]:
document_chunks[-2].page_content

'Y\nYaws 1266-1267\nforest 1379\nY chromosome 3373 (see also Genetic)\nabnormalities of 3005\nYeast infection (see also Fungal infection)\nvaginal 2542, 2544, 2545\nYellow fever 1400, 1429, 1437\nhepatic inflammation in 248\nvaccine against 1172, 1437, 3441\nYellow nail syndrome 732, 1995\npleural effusion in 1997\nYellow skin (see Jaundice)\nYersinia infection 1167, 1256-1257\nY. enterocolitica infection 147\nY. pestis infection 1924\nYew poisoning 3338\nYips 1762\nYo, antibodies to 1056\nYolk sac tumor 2476\nThe Merck Manual of Diagnosis & Therapy, 19th Edition\nY\n4103\nbhuvaneshwari_s@yahoo.com\nAT5WEX37OM\nThis file is meant for personal use by bhuvaneshwari_s@yahoo.com only.\nSharing or publishing the contents in part or full is liable for legal action.'

In [ ]:
document_chunks[-1].page_content

"Z\nZafirlukast 1879\nZalcitabine 1451\nin children 2854\nZaleplon 1709\nZanamivir 1407\nin influenza 1407, 1929\nZAP-70 (zeta-associated protein 70) deficiency 1092, 1108\nZavanelli maneuver 2680\nZellweger syndrome 2383, 3023\nZenker's diverticulum 125\nZidovudine 1451, 1453\nin children 2854\nZileuton 1881\nin asthma 1880\nZinc 49, 55, 3431-3432\nin common cold 1405\ndeficiency of 11, 49, 55\nin dermatophytoses 705\npoisoning with 3328, 3353\nrecommended dietary allowances for 50\nreference values for 3499\ntoxicity of 49, 55\ncopper deficiency and 49\nin Wilson's disease 52\nZinc oxide 2233\ngelatin formulation of 646, 672\nZinc pyrithione 647\nZinc shakes 55\nZipper injury 3239, 3240\nZiprasidone\nin agitation 1492\nin bipolar disorder 3059\npoisoning with 3347\nin schizophrenia 1566\nZoledronate 359, 361, 848\nZollinger-Ellison syndrome 95, 199, 200-201, 910\nmastocytosis vs 1125\nMenetrier's disease vs 132\npeptic ulcer disease vs 134\nZolmitriptan 1721\nZolpidem 1709, 3103\nZon

#### Observations - We can see the  contents of different pages from the Merck's manual.

### Embedding


A good method to choose an embedding model is to use the [embedding leaderboard](https://huggingface.co/spaces/mteb/leaderboard) and select an open-source model that is performant for the use case at hand.  Here, we are choosing general-purpose embedding model which is [`gte-large`](https://huggingface.co/thenlper/gte-large). The main reason for choosing this model is because of its embedding vector size, which is 512, the same as our token size in chunking.

In [23]:
embedding_model = SentenceTransformerEmbeddings(model_name="thenlper/gte-large")

/tmp/ipykernel_6157/3102610685.py:1: LangChainDeprecationWarning: The class `HuggingFaceEmbeddings` was deprecated in LangChain 0.2.2 and will be removed in 1.0. An updated version of the class exists in the :class:`~langchain-huggingface package and should be used instead. To use it run `pip install -U :class:`~langchain-huggingface` and import as `from :class:`~langchain_huggingface import HuggingFaceEmbeddings``.
  embedding_model = SentenceTransformerEmbeddings(model_name="thenlper/gte-large")


modules.json:   0%|          | 0.00/385 [00:00<?, ?B/s]

README.md: 0.00B [00:00, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/57.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/619 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/670M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/342 [00:00<?, ?B/s]

vocab.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

special_tokens_map.json:   0%|          | 0.00/125 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/191 [00:00<?, ?B/s]

In [24]:
embedding_1 = embedding_model.embed_query(document_chunks[0].page_content)
embedding_2 = embedding_model.embed_query(document_chunks[1].page_content)

In [25]:
print("Dimension of the embedding vector ",len(embedding_1))
len(embedding_1)==len(embedding_2)

Dimension of the embedding vector  1024


True

In [ ]:
embedding_1,embedding_2

([-0.027524471282958984,
  0.009962311014533043,
  -0.0017500261310487986,
  -0.010097301565110683,
  -0.017494982108473778,
  -0.02144297957420349,
  0.017750702798366547,
  0.036832600831985474,
  0.02047976478934288,
  0.021415451541543007,
  0.01699027605354786,
  0.00190501450560987,
  0.03343681991100311,
  -0.03891542926430702,
  -0.009305164217948914,
  -0.013668196275830269,
  -0.009174426086246967,
  -0.0387953445315361,
  0.0013768384233117104,
  0.02208181843161583,
  -0.0041916691698133945,
  0.020979003980755806,
  -0.070787213742733,
  -0.01839986816048622,
  0.011700845323503017,
  0.032903145998716354,
  0.015158338472247124,
  -0.013549075461924076,
  0.06471714377403259,
  0.05375861003994942,
  -0.037362899631261826,
  -0.006075256038457155,
  0.028080975636839867,
  -0.05745545029640198,
  -0.0202643983066082,
  -0.008325360715389252,
  0.043435387313365936,
  -0.03314726799726486,
  -0.011097319424152374,
  -0.03569364547729492,
  -0.0019112821901217103,
  -0.0050

We can see the Vector Embedding representations of the first 2 document chunks above.

### Vector Database

Creating a vector store from documents and embedding model specified earlier - (SentenceTransformerEmbeddings)The vector store data will be stored in a directory named medical_db.

In [26]:
out_dir = 'medical_db'

if not os.path.exists(out_dir):
  os.makedirs(out_dir)

In [27]:
vectorstore = Chroma.from_documents(
    document_chunks,
    embedding_model,
    persist_directory=out_dir
)

In [28]:
vectorstore = Chroma(persist_directory=out_dir,embedding_function=embedding_model)

/tmp/ipykernel_6157/2756559696.py:1: LangChainDeprecationWarning: The class `Chroma` was deprecated in LangChain 0.2.9 and will be removed in 1.0. An updated version of the class exists in the :class:`~langchain-chroma package and should be used instead. To use it run `pip install -U :class:`~langchain-chroma` and import as `from :class:`~langchain_chroma import Chroma``.
  vectorstore = Chroma(persist_directory=out_dir,embedding_function=embedding_model)


In [29]:
vectorstore.embeddings

HuggingFaceEmbeddings(client=SentenceTransformer(
  (0): Transformer({'max_seq_length': 512, 'do_lower_case': False, 'architecture': 'BertModel'})
  (1): Pooling({'word_embedding_dimension': 1024, 'pooling_mode_cls_token': False, 'pooling_mode_mean_tokens': True, 'pooling_mode_max_tokens': False, 'pooling_mode_mean_sqrt_len_tokens': False, 'pooling_mode_weightedmean_tokens': False, 'pooling_mode_lasttoken': False, 'include_prompt': True})
  (2): Normalize()
), model_name='thenlper/gte-large', cache_folder=None, model_kwargs={}, encode_kwargs={}, multi_process=False, show_progress=False)

In [30]:
vectorstore.similarity_search("What are the benefits of Magnesium in the body?",k=4) #Complete the code to pass a query and an appropriate k value

[Document(metadata={'creator': 'Atop CHM to PDF Converter', 'producer': 'pdf-lib (https://github.com/Hopding/pdf-lib)', 'moddate': '2026-05-02T04:34:41+00:00', 'creationDate': 'D:20120615054440Z', 'subject': '', 'title': 'The Merck Manual of Diagnosis & Therapy, 19th Edition', 'file_path': '/content/drive/MyDrive/GL/NLP-GenAI/Project/Project/medical_diagnosis_manual.pdf', 'keywords': '', 'author': '', 'modDate': 'D:20260502043441Z', 'source': '/content/drive/MyDrive/GL/NLP-GenAI/Project/Project/medical_diagnosis_manual.pdf', 'format': 'PDF 1.7', 'trapped': '', 'creationdate': '2012-06-15T05:44:40+00:00', 'page': 993, 'total_pages': 4114}, page_content="thiamine pyrophosphate cofactor activity and seems to stabilize the structure of macromolecules such as\nDNA and RNA. Mg is also related to Ca and K metabolism in an intimate but poorly understood way.\nHypomagnesemia\nHypomagnesemia is serum Mg concentration < 1.4 mEq/L (< 0.70 mmol/L). Causes include\ninadequate Mg intake and absorptio

We can see some similar documents retrieved for the question in the vectorstore similarity search for the benefits of Magnesium.

### Retriever

In [31]:
retriever = vectorstore.as_retriever(
    search_type='similarity',
    search_kwargs={'k': 2}
)

In [32]:
rel_docs = retriever.get_relevant_documents("What are the benefits of Magnesium in the body?")
rel_docs

/tmp/ipykernel_6157/3666318906.py:1: LangChainDeprecationWarning: The method `BaseRetriever.get_relevant_documents` was deprecated in langchain-core 0.1.46 and will be removed in 1.0. Use :meth:`~invoke` instead.
  rel_docs = retriever.get_relevant_documents("What are the benefits of Magnesium in the body?")


[Document(metadata={'subject': '', 'creationdate': '2012-06-15T05:44:40+00:00', 'title': 'The Merck Manual of Diagnosis & Therapy, 19th Edition', 'creator': 'Atop CHM to PDF Converter', 'author': '', 'keywords': '', 'page': 993, 'producer': 'pdf-lib (https://github.com/Hopding/pdf-lib)', 'trapped': '', 'source': '/content/drive/MyDrive/GL/NLP-GenAI/Project/Project/medical_diagnosis_manual.pdf', 'file_path': '/content/drive/MyDrive/GL/NLP-GenAI/Project/Project/medical_diagnosis_manual.pdf', 'moddate': '2026-05-02T04:34:41+00:00', 'creationDate': 'D:20120615054440Z', 'modDate': 'D:20260502043441Z', 'total_pages': 4114, 'format': 'PDF 1.7'}, page_content="thiamine pyrophosphate cofactor activity and seems to stabilize the structure of macromolecules such as\nDNA and RNA. Mg is also related to Ca and K metabolism in an intimate but poorly understood way.\nHypomagnesemia\nHypomagnesemia is serum Mg concentration < 1.4 mEq/L (< 0.70 mmol/L). Causes include\ninadequate Mg intake and absorptio

In [33]:
model_output = lcpp_llm(
      "What are the benefits of Magnesium in the body?",
      max_tokens=128,
      temperature=0,
    )

Llama.generate: prefix-match hit


In [34]:
model_output['choices'][0]['text']

'\nMagnesium is an essential mineral that plays a crucial role in many bodily functions. Some of the key benefits of magnesium in the body include:\n\n1. Bone health: Magnesium helps to build and maintain strong bones by regulating the metabolism of vitamin D and calcium. It also helps to stimulate the production of new bone tissue.\n2. Heart health: Magnesium helps to regulate heart rhythm, reduce blood pressure, and prevent blood clots. It also helps to reduce the risk of heart disease by reducing inflammation and'

The above response is generic and is solely based on the data the model was trained on, rather than the medical manual.  

Let's now provide our own context.

### System and User Prompt Template

Prompts guide the model to generate accurate responses. Here, we define two parts:

    1. The system message describing the assistant's role.
    2. A user message template including context and the question.

In [39]:
qna_system_message = """
You are an assistant whose work is to review the report and provide the appropriate answers from the context.
User input will have the context required by you to answer user questions.
This context will begin with the token: ###Context.
The context contains references to specific portions of a document relevant to the user query.

User questions will begin with the token: ###Question.

Please answer only using the context provided in the input. Do not mention anything about the context in your final answer.

If the answer is not found in the context, respond "I don't know".
"""

In [40]:
qna_user_message_template = """
Use the following context to answer the question.

Context:
{context}

Question:
{question}

Answer:
"""

### Response Function

Defining a function that will take a user query and LLM as an input, take the returned chunks as a context, and then use that context to generate an answer for the user query.

In [55]:
def generate_rag_response(user_input,k=3,max_tokens=256,temperature=0,top_p=0.95,top_k=50):
    global qna_system_message,qna_user_message_template
    # Retrieve relevant document chunks
    relevant_document_chunks = retriever.get_relevant_documents(query=user_input,k=k)
    context_list = [d.page_content for d in relevant_document_chunks]

    # Combine document chunks into a single context
    context_for_query = ". ".join(context_list)

    user_message = qna_user_message_template.replace('{context}', context_for_query)
    user_message = user_message.replace('{question}', user_input)

    prompt = qna_system_message + '\n' + user_message

    # Generate the response
    try:
        response = lcpp_llm(
                  prompt=prompt,
                  max_tokens=max_tokens,
                  temperature=temperature,
                  top_p=top_p,
                  top_k=top_k
                  )

        # Extract and print the model's response
        response = response['choices'][0]['text'].strip()
    except Exception as e:
        response = f'Sorry, I encountered the following error: \n {e}'

    return response

## Question Answering using RAG

### Query 1: What is the protocol for managing sepsis in a critical care unit?

In [56]:
user_input_1 = "What is the protocol for managing sepsis in a critical care unit?"
generate_rag_response(user_input_1,top_k=2)

Llama.generate: prefix-match hit


"According to the context, the protocol for managing sepsis in a critical care unit includes the following steps:\n\n1. Early recognition of sepsis and prompt administration of antibiotics.\n2. Use of point-of-care testing for blood tests, including electrolytes, CBC, and cultures.\n3. Monitoring of vital signs, including temperature, BP, pulse, and respiration rate.\n4. Measurement of fluid intake and output, and daily weight measurement.\n5. Use of complex devices to monitor physiologic parameters, with strict protocols for investigating alarms.\n6. Administration of antibiotics based on the suspected source of infection and the patient's clinical setting.\n7. Addition of vancomycin if resistant staphylococci or enterococci are suspected.\n8. Surgical drainage of abscesses and excision of necrotic tissues.\n9. Normalization of blood glucose levels with continuous IV insulin infusion.\n10. Corticosteroid therapy for patients with severe sepsis and"

### Query 2: What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?

In [90]:
user_input_2 = "What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?"
generate_rag_response(user_input_2,top_k=2,k=5)

Llama.generate: prefix-match hit


'The classic symptoms of acute appendicitis include epigastric or periumbilical pain followed by brief nausea, vomiting, and anorexia. However, these symptoms may not always be present, and the signs of tenderness in the right lower quadrant are not always obvious. Appendicitis is thought to result from obstruction of the appendiceal lumen, leading to distention, bacterial overgrowth, ischemia, and inflammation. Treatment is primarily surgical, with appendectomy being the most common procedure. In some cases, antibiotics may be used to treat the infection, but they are not a substitute for surgery. The surgical procedure of choice is usually a laparoscopic appendectomy, which has been shown to have fewer complications and a faster recovery time compared to open appendectomy.'

### Query 3: What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?

In [60]:
user_input_3 = "What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?"
generate_rag_response(user_input_3,top_k=15)

Llama.generate: prefix-match hit


'Sudden patchy hair loss, also known as alopecia areata, can be treated with topical corticosteroids, oral anti-inflammatory medications, or immunomodulatory drugs. The most effective treatment is thought to be intralesional triamcinolone acetonide injections. Other possible treatments include topical minoxidil and anthralin, as well as systemic corticosteroids. The exact cause of alopecia areata is not known, but it is believed to be an autoimmune disorder.\n\nPlease note that the answer provided is based on the information available in the given context and may not be applicable to all cases of sudden patchy hair loss. It is always recommended to consult a medical professional for proper diagnosis and treatment.'

### Query 4:  What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?

In [63]:
user_input_4 = "What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?"
generate_rag_response(user_input_4,top_k=4)

Llama.generate: prefix-match hit


"Based on the context provided, the recommended treatments for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function, are as follows:\n\n1. Initial treatment consists of ensuring a reliable airway and maintaining adequate ventilation, oxygenation, and blood pressure.\n2. Surgery is often needed in patients with more severe injury to place monitors to track and treat intracranial pressure, decompress the brain if intracranial pressure is increased, or remove intracranial hematomas.\n3. In the first few days after the injury, maintaining adequate brain perfusion and oxygenation and preventing complications of altered sensorium are important.\n4. Long-term treatment may involve a combination of drugs to control seizures and other symptoms, tailored to the patient's tolerance and response to the medications. The goal is to find the lowest effective dose that controls seizures with minimal adverse effects."

### Query 5: What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?

In [64]:
user_input_5 = "What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?"
generate_rag_response(user_input_5,top_k=3)

Llama.generate: prefix-match hit


'For a person with a fractured leg sustained during a hiking trip, it is essential to immobilize the affected limb using a splint or cast to prevent further injury and promote healing. The injured area should be elevated above heart level for the first 48 hours to minimize swelling. Pain management with analgesics is crucial, and periodic application of warmth may relieve pain and speed up the healing process. Prolonged immobilization can cause stiffness, contractures, and muscle atrophy, so early mobilization within the first few days or weeks may be necessary to minimize these complications. Infection should be monitored for, and if an odor emanates from within the cast or a fever develops, medical attention should be sought immediately. Good hygiene practices should be maintained throughout the recovery process.'

#### Observation - The responses for all the 5 queries are based on the context provided retrieved from the Merck's manual according to the context given through RAG model.  Let's see if fine-tuning the RAG model will provide better results.

### Fine-tuning

#### Query 1: What is the protocol for managing sepsis in a critical care unit?

In [65]:
user_input_1 = "What is the protocol for managing sepsis in a critical care unit?"
generate_rag_response(user_input_1,temperature=0.5,max_tokens=1024)

Llama.generate: prefix-match hit


'According to the context, the protocol for managing sepsis in a critical care unit includes the following steps:\n\n1. Early recognition of sepsis and prompt initiation of antibiotics.\n2. Use of point-of-care testing for blood tests, including electrolytes, CBC, and cultures.\n3. Management of fluid and electrolyte balance.\n4. Use of vasopressors to maintain mean arterial pressure ≥ 65 mmHg.\n5. Administration of corticosteroids.\n6. Monitoring for signs of organ dysfunction, such as acute respiratory distress syndrome (ARDS), acute kidney injury (AKI), and cardiovascular instability.\n7. Protocol-driven management of sepsis, including use of antibiotics, source control, and supportive care.\n8. Close monitoring of vital signs, laboratory values, and clinical status.\n9. Early recognition and management of complications such as abscesses, necrotic tissues, and bleeding.\n10. Use of invasive and non-invasive ventilation as needed.'

######Observations - Increasing the max_tokens to 1024  yields similar response in terms of token limits after fine-tuning for the above query, but has given complete sentences unlike the truncation we see in the earlier RAG retrivel prior to fine tuning. Answer seems more structured and coherent in terms of redefining certain protocols  after fine-tuning!

#### Query 2: What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?

In [91]:
user_input_2 = "What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?"
generate_rag_response(user_input_2,temperature=0.3,max_tokens=512,k=5,top_k=2)

Llama.generate: prefix-match hit


'The classic symptoms of acute appendicitis include epigastric or periumbilical pain followed by brief nausea, vomiting, and anorexia. However, these symptoms may not always be present, and the signs are not specific. Imaging studies should be done without delay in patients with atypical or equivocal findings. Appendicitis is primarily a clinical diagnosis, and early surgery is essential for a good prognosis. Antibiotics have no role in the treatment of appendicitis, and appendectomy is the only effective curative procedure. The surgical procedure for treating appendicitis is an open or laparoscopic appendectomy.'

###### Observations - The response is certainly very precise and concise  in retrieved context compared to response prior to fine-tuning. Follows the RAG instructions better and directly answers the question. Good retrieval output.

#### Query 3: What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?

In [67]:
user_input_3 = "What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?"
generate_rag_response(user_input_3,temperature=0.6,max_tokens=512)


Llama.generate: prefix-match hit


'Sudden patchy hair loss, also known as alopecia areata, can be treated with topical corticosteroids, oral anti-inflammatory medications, or immunomodulators. The most effective treatment is thought to be topical minoxidil, which prolongs the anagen growth phase and gradually enlarges miniaturized follicles into mature terminal hairs. However, only 30% to 40% of patients experience significant hair growth. Finasteride can also be used to treat male-pattern hair loss by inhibiting the conversion of testosterone to dihydrotestosterone. The possible causes of sudden patchy hair loss include autoimmune disorders, fungal infections, and physical trauma to the scalp. A thorough evaluation is necessary to determine the underlying cause and develop an appropriate treatment plan.'

###### Observations - The response is more random  after fine tuning as we have increased the temperature compared to the earlier RAG query retrieval with a lower temperature. In the statement " The most effective treatment is thought to be.." feels like hallucinating , where as the earlier response was very precise and relevant to the context question!

#### Query 4:  What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?

In [68]:
user_input_4 = "What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?"
generate_rag_response(user_input_4,temperature=0.1,max_tokens=824)

Llama.generate: prefix-match hit


"Based on the context provided, the recommended treatments for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function, are as follows:\n\n1. Initial treatment consists of ensuring a reliable airway and maintaining adequate ventilation, oxygenation, and blood pressure.\n2. Surgery is often needed in patients with more severe injury to place monitors to track and treat intracranial pressure, decompress the brain if intracranial pressure is increased, or remove intracranial hematomas.\n3. In the first few days after the injury, maintaining adequate brain perfusion and oxygenation and preventing complications of altered sensorium are important.\n4. Long-term treatment may involve a combination of drugs to control seizures and other symptoms, tailored to the patient's tolerance and response to the medications. The goal is to find the lowest effective dose that controls seizures and minimizes adverse effects."

###### Observations - Both the responses looks similar in terms of the context and the number of tokens generated in the responses prior to and after fine tuning (inspite of increasing the max_tokens to 824). This could be due to just a 0.1 increase in the temperature that it remained more predictive to the context given.  

#### Query 5: What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?

In [69]:
user_input_5 = "What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?"
generate_rag_response(user_input_5,temperature=0.5,max_tokens=656)


Llama.generate: prefix-match hit


'Fractures of the leg can be stabilized with immobilization in a cast or splint. To prevent compartment syndrome, the cast or splint should be cut open from end to end medially and laterally if swelling is severe enough to contribute to compartment syndrome. The patient should be advised to keep the cast dry, never put an object inside the cast, inspect the edges of the cast every day, and seek medical care if they notice any redness or soreness around the cast edges. Additionally, they should avoid putting weight on the affected leg and apply ice to reduce pain and swelling. Early mobilization can help prevent contractures and muscle atrophy, which may accelerate functional recovery.'

###### Observations - Fine tuned retrieval is concise and specific  in terms of clinical response to the query ( treatment , precautions and care required )

## Output Evaluation

Let us now use the LLM-as-a-judge method to check the quality of the RAG system on two parameters - retrieval and generation. We illustrate this evaluation based on the answers generated to the question from the previous section.

The two variables, groundedness_rater_system_message, and relevance_rater_system_message, define the evaluation criteria for assessing the quality of the generated answers.

In [76]:
groundedness_rater_system_message = """
You are tasked with rating AI generated answers to questions posed by users.
You will be presented a question, context used by the AI system to generate the answer and an AI generated answer to the question.
In the input, the question will begin with ###Question, the context will begin with ###Context while the AI generated answer will begin with ###Answer.

Evaluation criteria:
The task is to judge the extent to which the metric is followed by the answer.
1 - The metric is not followed at all
2 - The metric is followed only to a limited extent
3 - The metric is followed to a good extent
4 - The metric is followed mostly
5 - The metric is followed completely

Metric:
The answer should be derived only from the information presented in the context

Instructions:
1. First write down the steps that are needed to evaluate the answer as per the metric.
2. Give a step-by-step explanation if the answer adheres to the metric considering the question and context as the input.
3. Next, evaluate the extent to which the metric is followed.
4. Use the previous information to rate the answer using the evaluaton criteria and assign a score.
"""

In [77]:
relevance_rater_system_message = """
You are tasked with rating AI generated answers to questions posed by users.
You will be presented a question, context used by the AI system to generate the answer and an AI generated answer to the question.
In the input, the question will begin with ###Question, the context will begin with ###Context while the AI generated answer will begin with ###Answer.

Evaluation criteria:
The task is to judge the extent to which the metric is followed by the answer.
1 - The metric is not followed at all
2 - The metric is followed only to a limited extent
3 - The metric is followed to a good extent
4 - The metric is followed mostly
5 - The metric is followed completely

Metric:
Relevance measures how well the answer addresses the main aspects of the question, based on the context.
Consider whether all and only the important aspects are contained in the answer when evaluating relevance.

Instructions:
1. First write down the steps that are needed to evaluate the context as per the metric.
2. Give a step-by-step explanation if the context adheres to the metric considering the question as the input.
3. Next, evaluate the extent to which the metric is followed.
4. Use the previous information to rate the context using the evaluaton criteria and assign a score.
"""

In [78]:
user_message_template = """
###Question
{question}

###Context
{context}

###Answer
{answer}
"""

In [79]:
def generate_ground_relevance_response(user_input,k=3,max_tokens=128,temperature=0,top_p=0.95,top_k=50):
    global qna_system_message,qna_user_message_template
    # Retrieve relevant document chunks
    relevant_document_chunks = retriever.get_relevant_documents(query=user_input,k=3)
    context_list = [d.page_content for d in relevant_document_chunks]
    context_for_query = ". ".join(context_list)

    # Combine user_prompt and system_message to create the prompt
    prompt = f"""[INST]{qna_system_message}\n
                {'user'}: {qna_user_message_template.format(context=context_for_query, question=user_input)}
                [/INST]"""

    response = lcpp_llm(
            prompt=prompt,
            max_tokens=max_tokens,
            temperature=temperature,
            top_p=top_p,
            top_k=top_k,
            stop=['INST'],
            )

    answer =  response["choices"][0]["text"]

    # Combine user_prompt and system_message to create the prompt
    groundedness_prompt = f"""[INST]{groundedness_rater_system_message}\n
                {'user'}: {user_message_template.format(context=context_for_query, question=user_input, answer=answer)}
                [/INST]"""

    # Combine user_prompt and system_message to create the prompt
    relevance_prompt = f"""[INST]{relevance_rater_system_message}\n
                {'user'}: {user_message_template.format(context=context_for_query, question=user_input, answer=answer)}
                [/INST]"""

    response_1 = lcpp_llm(
            prompt=groundedness_prompt,
            max_tokens=max_tokens,
            temperature=temperature,
            top_p=top_p,
            top_k=top_k,
            stop=['INST'],
            )

    response_2 = lcpp_llm(
            prompt=relevance_prompt,
            max_tokens=max_tokens,
            temperature=temperature,
            top_p=top_p,
            top_k=top_k,
            stop=['INST'],
            )

    return response_1['choices'][0]['text'],response_2['choices'][0]['text']

### Query 1: What is the protocol for managing sepsis in a critical care unit?

In [80]:
ground,rel = generate_ground_relevance_response(user_input="What is the protocol for managing sepsis in a critical care unit?",max_tokens=370)

print(ground,end="\n\n")
print(rel)

Llama.generate: prefix-match hit
Llama.generate: prefix-match hit
Llama.generate: prefix-match hit


  Sure! I can help you with that. Here are the steps to evaluate the answer as per the metric:

Step 1: Check if the answer is derived only from the information presented in the context.

The answer provides a detailed protocol for managing sepsis in a critical care unit, which is based on the information provided in the context. Therefore, the first step is met.

Step 2: Evaluate the extent to which the metric is followed.

The answer follows the metric completely as it only provides information that is derived from the context and does not include any external or personal opinions.

Step 3: Use the previous information to rate the answer using the evaluation criteria.

Based on the previous steps, I would rate the answer as 5 - The metric is followed completely.

The answer provides a detailed protocol for managing sepsis in a critical care unit, which is derived only from the information presented in the context. Therefore, the answer fully meets the metric and the evaluation criter

   

*   As specified in the response above, the model answers the specific protocols for managing sepsis very clearly and it addresses all the aspects of the question with relevance .




### Query 2: What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?

In [92]:
ground,rel = generate_ground_relevance_response(user_input="What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?",max_tokens=1024)

print(ground,end="\n\n")
print(rel)

Llama.generate: prefix-match hit
Llama.generate: prefix-match hit
Llama.generate: prefix-match hit


  Sure, I can help you rate the answer based on the evaluation criteria provided. Here are the steps to evaluate the answer:

Step 1: Identify the metric - The metric is "The answer should be derived only from the information presented in the context."

Step 2: Evaluate the answer - The answer provides a comprehensive list of common symptoms of appendicitis and the recommended surgical procedure for treating it. The answer is derived solely from the information presented in the context.

Step 3: Rate the extent to which the metric is followed - The answer fully adheres to the metric as it only provides information that is present in the context. Therefore, I would rate the extent to which the metric is followed as "5 - The metric is followed completely."

Based on the previous information, I would assign a score of 5 out of 5 to the answer.

  Sure, I can help you with that! Here are the steps to evaluate the context as per the metric:

Step 1: Identify the main aspects of the question



*  The evaluation suggests that the answers were
  partially addressed for the second part of the question regarding the medicinal cure and surgical procedure treatments. It has correctly answered the symptoms of appendicitis , hence the metric score for the first part is 5 but did not give the full score for the second part of the question as there were loopholes in not answering clearly with groundedness , hence metric score was given as 2.

### Query 3: What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?

In [85]:
ground,rel = generate_ground_relevance_response(user_input="What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?",max_tokens=512)

print(ground,end="\n\n")
print(rel)

Llama.generate: prefix-match hit
Llama.generate: prefix-match hit
Llama.generate: prefix-match hit


  Sure! Here are the steps to evaluate the answer based on the metric provided:

Step 1: Identify the information presented in the context that the answer should be derived from.

The context provides information about the causes and treatments for sudden patchy hair loss, including alopecia areata, androgenetic alopecia, traction alopecia, tinea capitis, and trichotillomania.

Step 2: Evaluate whether the answer is derived only from the information presented in the context.

The answer provides a list of effective treatments for sudden patchy hair loss, including minoxidil, finasteride, and surgical options. These treatments are all mentioned in the context as potential treatment options for the causes of hair loss listed. Therefore, the answer is derived only from the information presented in the context.

Step 3: Evaluate the extent to which the metric is followed by the answer.

The answer provides a list of treatments that are all supported by evidence from the context. The answer


* Better fine-tuning the parameters for this query would result in good metric scores. The RAG model was hallucinating after increasing the temperature in the fine-tuning. The response was not concise and relevant enough, hence obtained low metric scores for relevance.

### Query 4: What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?

In [86]:
ground,rel = generate_ground_relevance_response(user_input="What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?",max_tokens=512)

print(ground,end="\n\n")
print(rel)

Llama.generate: prefix-match hit
Llama.generate: prefix-match hit
Llama.generate: prefix-match hit


  Sure, I can help you rate the answer based on the evaluation criteria provided. Here are the steps to evaluate the answer:

Step 1: Identify the metric - The metric is "The answer should be derived only from the information presented in the context."

Step 2: Evaluate the answer - The answer provides a list of recommended treatments for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function. The answer is based on the information presented in the context, which includes the causes of traumatic brain injury, the types of injuries that can occur, and the principles of long-term treatment.

Step 3: Rate the extent to which the metric is followed - Based on the information provided in the answer, it appears that the answer is derived entirely from the information presented in the context. The answer does not introduce any new or external information that is not supported by the context. Therefore, I would rate the ex

* The answer adheres to the context very well and completely relevant hence obtained the maximum metric scores. It has the lowest temperature so more predictive and precise to the context given.   

### Query 5: What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?

In [87]:
ground,rel = generate_ground_relevance_response(user_input="What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?",max_tokens=1024)

print(ground,end="\n\n")
print(rel)

Llama.generate: prefix-match hit
Llama.generate: prefix-match hit
Llama.generate: prefix-match hit


  Sure, I can help you with that! To evaluate the answer, we need to follow the metric provided in the instructions: "The answer should be derived only from the information presented in the context."

Let's go through the answer step by step and evaluate its adherence to the metric:

1. Immobilization: The answer mentions immobilization as one of the necessary precautions for a person who has fractured their leg during a hiking trip. This is correct, as immobilization is essential to prevent further injury and promote healing.

Adherence to the metric: Fully adheres

2. Pain management: The answer also mentions pain management as an important aspect of care for a person with a fractured leg. This is correct, as pain management is crucial to ensure the patient's comfort and mobility during the healing process.

Adherence to the metric: Fully adheres

3. Elevation: The answer suggests elevating the injured limb above the heart level to reduce swelling and promote blood flow. This is corr


* From the detailed response above, model answers completely relevant to the question and the metric is fully followed. This could be due to a very common and widely generic question asked. It covers all aspects of injury pertinent to the question asked hence with complete relevance and groundedness to the context !!

## Actionable Insights and Business Recommendations

- Prompt engineering with the appropriate,clear and specific prompts gave very good answers  based of LLM retrieval. Experimenting with more efficient prompts based of medical queries and analysis would help in further relevant answers!

- Retrieval parameter **`k`** is critical as the answer can be spread across multiple contexts. Query on appendicitis symptoms could not retrieve information when k was less than 5. It had to be increased to 5 to give a response. Further refinement of the K parameter for the queries would yield better responses with relevance and faithfulness to the context overall!

- **`chunk_overlap`** can be increased as the manual seems to have several spaces and common page endings for every page. It further ensures coherence, especially when context spans across chunks.
- **`max_tokens`** depends on query complexity; higher values yield detailed responses, while simple queries result in concise outputs despite large token limits due to prompt design and zero **`temperature`**.
- Refine prompt design and temperature settings to control response length and creativity.
- Continuously adjust RAG parameters based on specific use cases for optimal performance.

- Prioritize groundedness and relevance in evaluations to ensure reliable and contextually accurate outputs.

- Establish a feedback loop to fine-tune parameters, improving performance for diverse query types.

- Experimenting with various available LLMs and embedding models which are potentially smarter in information retrieval specific to medical diagnosis should be considered vital for model improvisations!

<font size=6 color='blue'>Power Ahead</font>
___